src/tools/pkcs11-tool.c

```diff
@@ -1238,6 +1291,94 @@ int main(int argc, char * argv[])
 	}
 
+	if (do_unwrap) {
+		CK_ULONG in_len;
+		unsigned char in_buffer[1024];
+		int r, fd;
+		CK_BYTE_PTR pWrappedKey;
+
+		if (opt_input == NULL)
+			fd = 0;
+		else if ((fd = open(opt_input, O_RDONLY | O_BINARY)) < 0)
+			util_fatal("Cannot open %s: %m", opt_input);
+
+		r = read(fd, in_buffer, sizeof(in_buffer));
+		if (r < 0)
+			util_fatal("Cannot read from %s: %m", opt_input);
+		in_len = r;
+		if (fd != 0)
+			close(fd);
+
+		pWrappedKey = (CK_BYTE_PTR) in_buffer;
+		CK_OBJECT_HANDLE hSecretKey;
+		if (!find_object(session, CKO_PRIVATE_KEY, &object,
+				 opt_object_id_len ? opt_object_id : NULL, opt_object_id_len, 0))
+			if (!find_object(session, CKO_SECRET_KEY, &object,
+					 opt_object_id_len ? opt_object_id : NULL, opt_object_id_len, 0))
+				util_fatal("Private/secret key not found");
+
+		unwrap_key(session, &hSecretKey, object, pWrappedKey, in_len, opt_key_type);
+	}
+
+	if (do_wrap) {
+		CK_BYTE pWrappedKey[4096];
+		CK_ULONG pulWrappedKeyLen = sizeof(pWrappedKey);
+		CK_MECHANISM mechanism;
+		CK_OBJECT_HANDLE hWrappingKey;	// wrapping key
+		CK_OBJECT_HANDLE hkey;	// key to be wrapped
+		CK_RV rv;
+		CK_BYTE hkey_id;
+		size_t hkey_id_len;
+		int fd, r;
+		CK_BYTE_PTR iv = NULL;
+		size_t iv_size = 0;
+
+		if (NULL == opt_application_id)
+			util_fatal("Use --application-id to specify secret key (to be wrapped)");
+		if (!opt_mechanism_used)
+			util_fatal("Unable to wrap, no mechanism specified\n");
+
+		mechanism.mechanism = opt_mechanism;
+		mechanism.pParameter = NULL_PTR;
+		mechanism.ulParameterLen = 0;
+		if (opt_mechanism == CKM_AES_CBC || opt_mechanism == CKM_AES_CBC_PAD) {
+			iv_size = 16;
+			iv = get_iv(opt_iv, &iv_size);
+			mechanism.pParameter = iv;
+			mechanism.ulParameterLen = iv_size;
+		}
+
+		hkey_id_len = sizeof(hkey_id);
+		if (!hex_to_bin(opt_application_id, &hkey_id, &hkey_id_len))
+			util_fatal("Invalid application-id \"%s\"\n", opt_application_id);
+
+		if (!find_object(session, CKO_SECRET_KEY, &hkey, hkey_id_len ? &hkey_id : NULL, hkey_id_len, 0))
+			util_fatal("Secret key (to be wrapped) not found");
+
+		if (!find_object(session, CKO_PUBLIC_KEY, &hWrappingKey,
+				 opt_object_id_len ? opt_object_id : NULL, opt_object_id_len, 0))
+			if (!find_object(session, CKO_SECRET_KEY, &hWrappingKey,
+					 opt_object_id_len ? opt_object_id : NULL, opt_object_id_len, 0))
+				util_fatal("Public/secret key (wrapping key) not found");
+
+		rv = p11->C_WrapKey(session, &mechanism, hWrappingKey, hkey, pWrappedKey, &pulWrappedKeyLen);
+		if (rv != CKR_OK)
+			p11_fatal("C_WrapKey", rv);
+		printf("Key wrapped\n");
+
+		if (opt_output == NULL)
+			fd = 1;
+		else if ((fd = open(opt_output, O_CREAT | O_TRUNC | O_WRONLY | O_BINARY, S_IRUSR | S_IWUSR)) < 0)
+			util_fatal("failed to open %s: %m", opt_output);
+
+		r = write(fd, pWrappedKey, pulWrappedKeyLen);
+
+		if (r < 0)
+			util_fatal("Failed to write to %s: %m", opt_output);
+		if (fd != 1)
+			close(fd);
+		free(iv);
+	}
```

Comment:
please, create a separate function for wrap_key. This function is already long enough.